mypy/test/testpep561.py

```diff
@@ -19,112 +22,112 @@
 
 
 def check_mypy_run(cmd_line: List[str],
-                   expected_out: str,
+                   python_executable: str = sys.executable,
+                   expected_out: str = '',
                    expected_err: str = '',
                    expected_returncode: int = 1) -> None:
     """Helper to run mypy and check the output."""
+    if python_executable != sys.executable:
+        cmd_line.append('--python-executable={}'.format(python_executable))
     out, err, returncode = mypy.api.run(cmd_line)
     assert out == expected_out, err
     assert err == expected_err, out
     assert returncode == expected_returncode, returncode
 
 
-def is_in_venv() -> bool:
-    """Returns whether we are running inside a venv.
-
-    Based on https://stackoverflow.com/a/42580137.
-
-    """
-    if hasattr(sys, 'real_prefix'):
-        return True
-    else:
-        return hasattr(sys, 'base_prefix') and sys.base_prefix != sys.prefix
+class TestPEP561(TestCase):
 
+    @contextmanager
+    def virtualenv(self, python_executable: str = sys.executable) -> Generator[str, None, None]:
+        """Context manager that creates a virtualenv in a temporary directory
+
+        returns the path to the created Python executable"""
+        with tempfile.TemporaryDirectory() as venv_dir:
+            run_command([sys.executable, '-m', 'virtualenv', '-p{}'.format(python_executable),
+                        venv_dir], cwd=os.getcwd())
+            if sys.platform == 'win32':
+                yield os.path.abspath(os.path.join(venv_dir, 'Scripts', 'python'))
+            else:
+                yield os.path.abspath(os.path.join(venv_dir, 'bin', 'python'))
 
-class TestPEP561(TestCase):
     @contextmanager
     def install_package(self, pkg: str,
                         python_executable: str = sys.executable) -> Iterator[None]:
         """Context manager to temporarily install a package from test-data/packages/pkg/"""
         working_dir = os.path.join(package_path, pkg)
         install_cmd = [python_executable, '-m', 'pip', 'install', '.']
-        # if we aren't in a virtualenv, install in the
-        # user package directory so we don't need sudo
-        if not is_in_venv() or python_executable != sys.executable:
-            install_cmd.append('--user')
         returncode, lines = run_command(install_cmd, cwd=working_dir)
         if returncode != 0:
             self.fail('\n'.join(lines))
-        try:
-            yield
-        finally:
-            returncode, lines = run_command([python_executable, '-m', 'pip', 'uninstall',
-                                             '-y', pkg], cwd=package_path)
-            if returncode != 0:
-                self.fail('\n'.join(lines))
+        yield
```

Comment:
I am not sure what is the benefit of context manager which has yield as a last statement. Can't be this just made a function?

Reply:
Good point. I will refactor things into just using this as a function.